#Task 3: Language model using the text from a large collection of Science Fiction stories

In [4]:
#importing the necessary libraries
from tensorflow.keras.callbacks import LambdaCallback
from tensorflow.keras.models import Sequential, load_model
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import LSTM, Embedding, GRU, Dropout
from tensorflow.keras.optimizers import RMSprop
from tensorflow.keras.utils import get_file, to_categorical
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.callbacks import LearningRateScheduler, ModelCheckpoint, EarlyStopping
import matplotlib.pyplot as plt
import numpy as np
import random
import sys
import io
import requests
import re
import string
from google.colab import drive
import zipfile

In [5]:
#mounting google drive
drive.mount('/content/drive', force_remount=True) 

Mounted at /content/drive


Data preparation

In [6]:
#loading the data
zip_file = "/content/drive/My Drive/Deep Learning/SciFi.zip"
# extract location
extract_path = "/content/DeepLearning/Scifi/"
# create a ZipFile object
with zipfile.ZipFile(zip_file, 'r') as zip_obj:
  # extract all the contents of the zip file to the specified directory 
  zip_obj.extractall(extract_path)

In [7]:
#open the extracted zip file
file=open('/content/DeepLearning/Scifi/internet_archive_scifi_v3.txt','r',encoding='utf8').read()

In [8]:
len(file)

149326361

In [9]:
print(file[:500])

MARCH # All Stories New and Complete Publisher Editor IF is published bi-monthly by Quinn Publishing Company, Inc., Kingston, New York. Volume #, No. #. Copyright # by Quinn Publishing Company, Inc. Application for Entry' as Second Class matter at Post Office, Buffalo, New York, pending. Subscription # for # issues in U.S. and Possessions: Canada # for # issues; elsewhere #. Aiiow four weeks for change of address. All stories appearing in this magazine are fiction. Any similarity to actual perso


Data Preprocessing

In [10]:
#Data cleaning - Splitting the text, Removing the special characters and extra space
def cleaned_text(text):
  text = text.split(" ")
  text = re.sub(r'[^\x00-\x7f]', r'', str(text)) # removing special chars...
  text = text.translate(str.maketrans('', '', string.punctuation)) # removing special chars...
  text = re.sub('\s+', ' ', str(text)) # removing extra spaces...
  return text

text = cleaned_text(file)

In [30]:
#visualize cleaned text
print("Example of cleaned text:",text[0:100])
print("Length of text:", len(text))

Example of cleaned text: MARCH All Stories New and Complete Publisher Editor IF is published bimonthly by Quinn Publishing Co
Length of text: 142442755


In [27]:
#creating a corpus for the text and creating a list of unique words
text_corpus = text.split(" ")
text_corpus = [x for x in text_corpus if x != ""]
print("Length of text corpus:",len(text_corpus))
unique_vocab = list(set(text_corpus))
print("Length of unique words:",len(unique_vocab))

Length of text corpus: 26308635
Length of unique words: 330125


In [16]:
text_corpus[:20]

['MARCH',
 'All',
 'Stories',
 'New',
 'and',
 'Complete',
 'Publisher',
 'Editor',
 'IF',
 'is',
 'published',
 'bimonthly',
 'by',
 'Quinn',
 'Publishing',
 'Company',
 'Inc',
 'Kingston',
 'New',
 'York']

In [26]:
#Dividing the whole text into multiple strings each of length 31
seq_length =  30
step_size = 1 
all_sentences = []
for i in range(seq_length, len(text_corpus)):
  # sliding window to divide the text
  sentence = text_corpus[i - seq_length: i] 
  sentence = ' '.join(sentence)
  all_sentences.append(sentence)

all_sentences[:10]
print("Length of all sentences: ", len(all_sentences))

Length of all sentences:  499970


In [24]:
# Data selected and unique vocabulary in the data
model_sent  = all_sentences[25808605:]
print(len(model_sent), "words are selected from the end as the smaller dataset")
print("Example of model sent: ", model_sent[:20])
unique_vocab = list(set(model_sent))
print("Length of unique words: ", len(unique_vocab))
print("Example of of unique words: ", unique_vocab[0:10])

500000 words are selected from the end as the smaller dataset
Example of model sent:  ['think He fell forward Or more he just crumpled to a heap on his Formula Hut floor and became right before her eyes a very sad pile of quite expensive', 'He fell forward Or more he just crumpled to a heap on his Formula Hut floor and became right before her eyes a very sad pile of quite expensive replacement', 'fell forward Or more he just crumpled to a heap on his Formula Hut floor and became right before her eyes a very sad pile of quite expensive replacement metal', 'forward Or more he just crumpled to a heap on his Formula Hut floor and became right before her eyes a very sad pile of quite expensive replacement metal and', 'Or more he just crumpled to a heap on his Formula Hut floor and became right before her eyes a very sad pile of quite expensive replacement metal and fleshstrip', 'more he just crumpled to a heap on his Formula Hut floor and became right before her eyes a very sad pile of quit

In [25]:
# tokenizing the words and converting to numerical values
tokenizer = Tokenizer(num_words= 10000)
tokenizer.fit_on_texts(model_sent)
seq = tokenizer.texts_to_sequences(model_sent)

In [32]:
# Pad sequences to a fixed length of 30 with 'pre'
max_length = 30
padded_sequences = pad_sequences(seq, maxlen=max_length, padding='pre', truncating='pre')

In [33]:
data = np.vstack(padded_sequences)

In [34]:
# selecting 30 columns of each rows as features and 31st as target variable
X = data[:, :-1]
y = data[:, -1]

In [35]:
len(tokenizer.word_index)

21544

In [ ]:
X.shape[1]

29

Model Building

In [36]:
# Sequential LSTM model to predict next word...
model = Sequential()

# input_dim is the length of the vocab/dictionary that we created earlier, output_dim is 50, and input length is 31...
model.add(Embedding(len(tokenizer.word_index) + 1, 50, input_length = X.shape[1])) 

# 64 LSTM units and return_sequences = True to pass it on to next LSTM layer. Dropout for regularization
model.add(LSTM(64, return_sequences=True))


model.add(LSTM(64))

model.add(Dense(128, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(len(tokenizer.word_index) + 1, activation='softmax'))
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 29, 50)            1077250   
                                                                 
 lstm (LSTM)                 (None, 29, 64)            29440     
                                                                 
 lstm_1 (LSTM)               (None, 64)                33024     
                                                                 
 dense (Dense)               (None, 128)               8320      
                                                                 
 dropout (Dropout)           (None, 128)               0         
                                                                 
 dense_1 (Dense)             (None, 21545)             2779305   
                                                                 
Total params: 3,927,339
Trainable params: 3,927,339
No

In [ ]:
#define early stopping and Model checkpoints that are used as callbacks in the model
es = EarlyStopping(monitor="loss",patience=10,verbose=0,mode="auto")
checkpoint_filepath = "/content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.{epoch:007}"
model_checkpoint_callback = ModelCheckpoint(
    filepath=checkpoint_filepath, monitor='loss', save_best_only=True, mode='min', verbose=1)


In [ ]:
# Compiling the model with adam optimizer and training it for 100 epochs with a batch size of 32
model.compile(loss='sparse_categorical_crossentropy', optimizer = 'adam', metrics = ['accuracy'])
model.fit(X, y, batch_size = 32, epochs=100, callbacks=[es,model_checkpoint_callback])
model.save('/content/drive/MyDrive/Deep Learning/task3/model_weights.h5')

Epoch 1/100
15625/15625 [==============================] - ETA: 0s - loss: 6.5371 - accuracy: 0.0819
Epoch 1: loss improved from inf to 6.53713, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000001


15625/15625 [==============================] - 160s 10ms/step - loss: 6.5371 - accuracy: 0.0819
Epoch 2/100
15622/15625 [============================>.] - ETA: 0s - loss: 6.0613 - accuracy: 0.1057
Epoch 2: loss improved from 6.53713 to 6.06130, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000002


15625/15625 [==============================] - 135s 9ms/step - loss: 6.0613 - accuracy: 0.1057
Epoch 3/100
15623/15625 [============================>.] - ETA: 0s - loss: 5.8554 - accuracy: 0.1155
Epoch 3: loss improved from 6.06130 to 5.85536, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000003


15625/15625 [==============================] - 134s 9ms/step - loss: 5.8554 - accuracy: 0.1155
Epoch 4/100
15625/15625 [==============================] - ETA: 0s - loss: 5.7063 - accuracy: 0.1226
Epoch 4: loss improved from 5.85536 to 5.70628, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000004


15625/15625 [==============================] - 135s 9ms/step - loss: 5.7063 - accuracy: 0.1226
Epoch 5/100
15625/15625 [==============================] - ETA: 0s - loss: 5.5919 - accuracy: 0.1284
Epoch 5: loss improved from 5.70628 to 5.59190, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000005


15625/15625 [==============================] - 133s 9ms/step - loss: 5.5919 - accuracy: 0.1284
Epoch 6/100
15624/15625 [============================>.] - ETA: 0s - loss: 5.4936 - accuracy: 0.1342
Epoch 6: loss improved from 5.59190 to 5.49353, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000006


15625/15625 [==============================] - 135s 9ms/step - loss: 5.4935 - accuracy: 0.1342
Epoch 7/100
15620/15625 [============================>.] - ETA: 0s - loss: 5.4041 - accuracy: 0.1392
Epoch 7: loss improved from 5.49353 to 5.40412, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000007


15625/15625 [==============================] - 134s 9ms/step - loss: 5.4041 - accuracy: 0.1392
Epoch 8/100
15623/15625 [============================>.] - ETA: 0s - loss: 5.3214 - accuracy: 0.1441
Epoch 8: loss improved from 5.40412 to 5.32133, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000008


15625/15625 [==============================] - 135s 9ms/step - loss: 5.3213 - accuracy: 0.1441
Epoch 9/100
15625/15625 [==============================] - ETA: 0s - loss: 5.2467 - accuracy: 0.1482
Epoch 9: loss improved from 5.32133 to 5.24669, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000009


15625/15625 [==============================] - 133s 9ms/step - loss: 5.2467 - accuracy: 0.1482
Epoch 10/100
15621/15625 [============================>.] - ETA: 0s - loss: 5.1768 - accuracy: 0.1519
Epoch 10: loss improved from 5.24669 to 5.17678, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000010


15625/15625 [==============================] - 136s 9ms/step - loss: 5.1768 - accuracy: 0.1519
Epoch 11/100
15621/15625 [============================>.] - ETA: 0s - loss: 5.1145 - accuracy: 0.1558
Epoch 11: loss improved from 5.17678 to 5.11457, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000011


15625/15625 [==============================] - 133s 9ms/step - loss: 5.1146 - accuracy: 0.1558
Epoch 12/100
15623/15625 [============================>.] - ETA: 0s - loss: 5.0561 - accuracy: 0.1593
Epoch 12: loss improved from 5.11457 to 5.05603, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000012


15625/15625 [==============================] - 133s 8ms/step - loss: 5.0560 - accuracy: 0.1593
Epoch 13/100
15624/15625 [============================>.] - ETA: 0s - loss: 5.0035 - accuracy: 0.1630
Epoch 13: loss improved from 5.05603 to 5.00358, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000013


15625/15625 [==============================] - 135s 9ms/step - loss: 5.0036 - accuracy: 0.1630
Epoch 14/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.9549 - accuracy: 0.1660
Epoch 14: loss improved from 5.00358 to 4.95495, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000014


15625/15625 [==============================] - 136s 9ms/step - loss: 4.9549 - accuracy: 0.1660
Epoch 15/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.9124 - accuracy: 0.1690
Epoch 15: loss improved from 4.95495 to 4.91243, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000015


15625/15625 [==============================] - 134s 9ms/step - loss: 4.9124 - accuracy: 0.1690
Epoch 16/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.8711 - accuracy: 0.1713
Epoch 16: loss improved from 4.91243 to 4.87120, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000016


15625/15625 [==============================] - 134s 9ms/step - loss: 4.8712 - accuracy: 0.1713
Epoch 17/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.8345 - accuracy: 0.1736
Epoch 17: loss improved from 4.87120 to 4.83444, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000017


15625/15625 [==============================] - 137s 9ms/step - loss: 4.8344 - accuracy: 0.1736
Epoch 18/100
15623/15625 [============================>.] - ETA: 0s - loss: 4.7971 - accuracy: 0.1763
Epoch 18: loss improved from 4.83444 to 4.79717, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000018


15625/15625 [==============================] - 135s 9ms/step - loss: 4.7972 - accuracy: 0.1763
Epoch 19/100
15625/15625 [==============================] - ETA: 0s - loss: 4.7664 - accuracy: 0.1784
Epoch 19: loss improved from 4.79717 to 4.76642, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000019


15625/15625 [==============================] - 134s 9ms/step - loss: 4.7664 - accuracy: 0.1784
Epoch 20/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.7365 - accuracy: 0.1806
Epoch 20: loss improved from 4.76642 to 4.73654, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000020


15625/15625 [==============================] - 134s 9ms/step - loss: 4.7365 - accuracy: 0.1806
Epoch 21/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.7087 - accuracy: 0.1827
Epoch 21: loss improved from 4.73654 to 4.70870, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000021


15625/15625 [==============================] - 136s 9ms/step - loss: 4.7087 - accuracy: 0.1827
Epoch 22/100
15624/15625 [============================>.] - ETA: 0s - loss: 4.6836 - accuracy: 0.1849
Epoch 22: loss improved from 4.70870 to 4.68362, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000022


15625/15625 [==============================] - 133s 8ms/step - loss: 4.6836 - accuracy: 0.1849
Epoch 23/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.6591 - accuracy: 0.1856
Epoch 23: loss improved from 4.68362 to 4.65918, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000023


15625/15625 [==============================] - 135s 9ms/step - loss: 4.6592 - accuracy: 0.1856
Epoch 24/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.6356 - accuracy: 0.1877
Epoch 24: loss improved from 4.65918 to 4.63554, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000024


15625/15625 [==============================] - 134s 9ms/step - loss: 4.6355 - accuracy: 0.1877
Epoch 25/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.6154 - accuracy: 0.1891
Epoch 25: loss improved from 4.63554 to 4.61531, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000025


15625/15625 [==============================] - 135s 9ms/step - loss: 4.6153 - accuracy: 0.1891
Epoch 26/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.5963 - accuracy: 0.1903
Epoch 26: loss improved from 4.61531 to 4.59632, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000026


15625/15625 [==============================] - 133s 9ms/step - loss: 4.5963 - accuracy: 0.1903
Epoch 27/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.5785 - accuracy: 0.1920
Epoch 27: loss improved from 4.59632 to 4.57853, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000027


15625/15625 [==============================] - 134s 9ms/step - loss: 4.5785 - accuracy: 0.1920
Epoch 28/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.5622 - accuracy: 0.1933
Epoch 28: loss improved from 4.57853 to 4.56215, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000028


15625/15625 [==============================] - 133s 8ms/step - loss: 4.5621 - accuracy: 0.1933
Epoch 29/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.5442 - accuracy: 0.1947
Epoch 29: loss improved from 4.56215 to 4.54425, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000029


15625/15625 [==============================] - 134s 9ms/step - loss: 4.5443 - accuracy: 0.1947
Epoch 30/100
15623/15625 [============================>.] - ETA: 0s - loss: 4.5312 - accuracy: 0.1956
Epoch 30: loss improved from 4.54425 to 4.53116, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000030


15625/15625 [==============================] - 135s 9ms/step - loss: 4.5312 - accuracy: 0.1956
Epoch 31/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.5163 - accuracy: 0.1970
Epoch 31: loss improved from 4.53116 to 4.51636, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000031


15625/15625 [==============================] - 133s 9ms/step - loss: 4.5164 - accuracy: 0.1970
Epoch 32/100
15624/15625 [============================>.] - ETA: 0s - loss: 4.5019 - accuracy: 0.1984
Epoch 32: loss improved from 4.51636 to 4.50194, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000032


15625/15625 [==============================] - 133s 9ms/step - loss: 4.5019 - accuracy: 0.1984
Epoch 33/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.4886 - accuracy: 0.1990
Epoch 33: loss improved from 4.50194 to 4.48860, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000033


15625/15625 [==============================] - 138s 9ms/step - loss: 4.4886 - accuracy: 0.1990
Epoch 34/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.4755 - accuracy: 0.1998
Epoch 34: loss improved from 4.48860 to 4.47548, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000034


15625/15625 [==============================] - 133s 9ms/step - loss: 4.4755 - accuracy: 0.1998
Epoch 35/100
15624/15625 [============================>.] - ETA: 0s - loss: 4.4610 - accuracy: 0.2009
Epoch 35: loss improved from 4.47548 to 4.46103, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000035


15625/15625 [==============================] - 136s 9ms/step - loss: 4.4610 - accuracy: 0.2009
Epoch 36/100
15625/15625 [==============================] - ETA: 0s - loss: 4.4486 - accuracy: 0.2022
Epoch 36: loss improved from 4.46103 to 4.44861, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000036


15625/15625 [==============================] - 134s 9ms/step - loss: 4.4486 - accuracy: 0.2022
Epoch 37/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.4391 - accuracy: 0.2025
Epoch 37: loss improved from 4.44861 to 4.43910, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000037


15625/15625 [==============================] - 136s 9ms/step - loss: 4.4391 - accuracy: 0.2025
Epoch 38/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.4281 - accuracy: 0.2037
Epoch 38: loss improved from 4.43910 to 4.42807, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000038


15625/15625 [==============================] - 136s 9ms/step - loss: 4.4281 - accuracy: 0.2037
Epoch 39/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.4188 - accuracy: 0.2047
Epoch 39: loss improved from 4.42807 to 4.41882, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000039


15625/15625 [==============================] - 134s 9ms/step - loss: 4.4188 - accuracy: 0.2047
Epoch 40/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.4089 - accuracy: 0.2053
Epoch 40: loss improved from 4.41882 to 4.40891, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000040


15625/15625 [==============================] - 133s 9ms/step - loss: 4.4089 - accuracy: 0.2053
Epoch 41/100
15624/15625 [============================>.] - ETA: 0s - loss: 4.3995 - accuracy: 0.2060
Epoch 41: loss improved from 4.40891 to 4.39955, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000041


15625/15625 [==============================] - 136s 9ms/step - loss: 4.3996 - accuracy: 0.2060
Epoch 42/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.3915 - accuracy: 0.2066
Epoch 42: loss improved from 4.39955 to 4.39145, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000042


15625/15625 [==============================] - 133s 9ms/step - loss: 4.3914 - accuracy: 0.2067
Epoch 43/100
15625/15625 [==============================] - ETA: 0s - loss: 4.3822 - accuracy: 0.2076
Epoch 43: loss improved from 4.39145 to 4.38215, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000043


15625/15625 [==============================] - 135s 9ms/step - loss: 4.3822 - accuracy: 0.2076
Epoch 44/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.3747 - accuracy: 0.2080
Epoch 44: loss improved from 4.38215 to 4.37470, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000044


15625/15625 [==============================] - 133s 9ms/step - loss: 4.3747 - accuracy: 0.2080
Epoch 45/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.3675 - accuracy: 0.2090
Epoch 45: loss improved from 4.37470 to 4.36745, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000045


15625/15625 [==============================] - 135s 9ms/step - loss: 4.3675 - accuracy: 0.2090
Epoch 46/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.3603 - accuracy: 0.2097
Epoch 46: loss improved from 4.36745 to 4.36028, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000046


15625/15625 [==============================] - 133s 9ms/step - loss: 4.3603 - accuracy: 0.2097
Epoch 47/100
15625/15625 [==============================] - ETA: 0s - loss: 4.3553 - accuracy: 0.2099
Epoch 47: loss improved from 4.36028 to 4.35534, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000047


15625/15625 [==============================] - 135s 9ms/step - loss: 4.3553 - accuracy: 0.2099
Epoch 48/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.3488 - accuracy: 0.2109
Epoch 48: loss improved from 4.35534 to 4.34883, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000048


15625/15625 [==============================] - 133s 9ms/step - loss: 4.3488 - accuracy: 0.2109
Epoch 49/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.3417 - accuracy: 0.2117
Epoch 49: loss improved from 4.34883 to 4.34187, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000049


15625/15625 [==============================] - 135s 9ms/step - loss: 4.3419 - accuracy: 0.2117
Epoch 50/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.3353 - accuracy: 0.2118
Epoch 50: loss improved from 4.34187 to 4.33529, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000050


15625/15625 [==============================] - 135s 9ms/step - loss: 4.3353 - accuracy: 0.2118
Epoch 51/100
15625/15625 [==============================] - ETA: 0s - loss: 4.3282 - accuracy: 0.2130
Epoch 51: loss improved from 4.33529 to 4.32824, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000051


15625/15625 [==============================] - 133s 9ms/step - loss: 4.3282 - accuracy: 0.2130
Epoch 52/100
15625/15625 [==============================] - ETA: 0s - loss: 4.3266 - accuracy: 0.2128
Epoch 52: loss improved from 4.32824 to 4.32657, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000052


15625/15625 [==============================] - 132s 8ms/step - loss: 4.3266 - accuracy: 0.2128
Epoch 53/100
15625/15625 [==============================] - ETA: 0s - loss: 4.3212 - accuracy: 0.2132
Epoch 53: loss improved from 4.32657 to 4.32119, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000053


15625/15625 [==============================] - 135s 9ms/step - loss: 4.3212 - accuracy: 0.2132
Epoch 54/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.3140 - accuracy: 0.2141
Epoch 54: loss improved from 4.32119 to 4.31391, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000054


15625/15625 [==============================] - 132s 8ms/step - loss: 4.3139 - accuracy: 0.2141
Epoch 55/100
15625/15625 [==============================] - ETA: 0s - loss: 4.3105 - accuracy: 0.2148
Epoch 55: loss improved from 4.31391 to 4.31047, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000055


15625/15625 [==============================] - 135s 9ms/step - loss: 4.3105 - accuracy: 0.2148
Epoch 56/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.3065 - accuracy: 0.2145
Epoch 56: loss improved from 4.31047 to 4.30657, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000056


15625/15625 [==============================] - 132s 8ms/step - loss: 4.3066 - accuracy: 0.2145
Epoch 57/100
15624/15625 [============================>.] - ETA: 0s - loss: 4.3019 - accuracy: 0.2147
Epoch 57: loss improved from 4.30657 to 4.30189, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000057


15625/15625 [==============================] - 135s 9ms/step - loss: 4.3019 - accuracy: 0.2146
Epoch 58/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.2969 - accuracy: 0.2159
Epoch 58: loss improved from 4.30189 to 4.29700, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000058


15625/15625 [==============================] - 133s 9ms/step - loss: 4.2970 - accuracy: 0.2159
Epoch 59/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.2925 - accuracy: 0.2156
Epoch 59: loss improved from 4.29700 to 4.29248, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000059


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2925 - accuracy: 0.2156
Epoch 60/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.2903 - accuracy: 0.2162
Epoch 60: loss improved from 4.29248 to 4.29037, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000060


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2904 - accuracy: 0.2162
Epoch 61/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.2839 - accuracy: 0.2173
Epoch 61: loss improved from 4.29037 to 4.28398, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000061


15625/15625 [==============================] - 133s 9ms/step - loss: 4.2840 - accuracy: 0.2173
Epoch 62/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.2792 - accuracy: 0.2174
Epoch 62: loss improved from 4.28398 to 4.27922, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000062


15625/15625 [==============================] - 133s 9ms/step - loss: 4.2792 - accuracy: 0.2173
Epoch 63/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.2765 - accuracy: 0.2176
Epoch 63: loss improved from 4.27922 to 4.27655, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000063


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2766 - accuracy: 0.2176
Epoch 64/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.2768 - accuracy: 0.2176
Epoch 64: loss did not improve from 4.27655
15625/15625 [==============================] - 123s 8ms/step - loss: 4.2769 - accuracy: 0.2176
Epoch 65/100
15625/15625 [==============================] - ETA: 0s - loss: 4.2706 - accuracy: 0.2181
Epoch 65: loss improved from 4.27655 to 4.27062, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000065


15625/15625 [==============================] - 132s 8ms/step - loss: 4.2706 - accuracy: 0.2181
Epoch 66/100
15624/15625 [============================>.] - ETA: 0s - loss: 4.2691 - accuracy: 0.2181
Epoch 66: loss improved from 4.27062 to 4.26912, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000066


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2691 - accuracy: 0.2181
Epoch 67/100
15623/15625 [============================>.] - ETA: 0s - loss: 4.2634 - accuracy: 0.2188
Epoch 67: loss improved from 4.26912 to 4.26335, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000067


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2634 - accuracy: 0.2188
Epoch 68/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.2600 - accuracy: 0.2189
Epoch 68: loss improved from 4.26335 to 4.25988, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000068


15625/15625 [==============================] - 133s 9ms/step - loss: 4.2599 - accuracy: 0.2190
Epoch 69/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.2592 - accuracy: 0.2194
Epoch 69: loss improved from 4.25988 to 4.25921, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000069


15625/15625 [==============================] - 133s 8ms/step - loss: 4.2592 - accuracy: 0.2194
Epoch 70/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.2546 - accuracy: 0.2195
Epoch 70: loss improved from 4.25921 to 4.25462, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000070


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2546 - accuracy: 0.2195
Epoch 71/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.2527 - accuracy: 0.2200
Epoch 71: loss improved from 4.25462 to 4.25261, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000071


15625/15625 [==============================] - 132s 8ms/step - loss: 4.2526 - accuracy: 0.2200
Epoch 72/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.2518 - accuracy: 0.2204
Epoch 72: loss improved from 4.25261 to 4.25172, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000072


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2517 - accuracy: 0.2204
Epoch 73/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.2459 - accuracy: 0.2205
Epoch 73: loss improved from 4.25172 to 4.24594, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000073


15625/15625 [==============================] - 133s 8ms/step - loss: 4.2459 - accuracy: 0.2205
Epoch 74/100
15623/15625 [============================>.] - ETA: 0s - loss: 4.2438 - accuracy: 0.2210
Epoch 74: loss improved from 4.24594 to 4.24377, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000074


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2438 - accuracy: 0.2210
Epoch 75/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.2417 - accuracy: 0.2214
Epoch 75: loss improved from 4.24377 to 4.24166, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000075


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2417 - accuracy: 0.2214
Epoch 76/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.2435 - accuracy: 0.2212
Epoch 76: loss did not improve from 4.24166
15625/15625 [==============================] - 123s 8ms/step - loss: 4.2435 - accuracy: 0.2212
Epoch 77/100
15624/15625 [============================>.] - ETA: 0s - loss: 4.2409 - accuracy: 0.2217
Epoch 77: loss improved from 4.24166 to 4.24091, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000077


15625/15625 [==============================] - 133s 9ms/step - loss: 4.2409 - accuracy: 0.2217
Epoch 78/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.2398 - accuracy: 0.2218
Epoch 78: loss improved from 4.24091 to 4.23977, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000078


15625/15625 [==============================] - 134s 9ms/step - loss: 4.2398 - accuracy: 0.2218
Epoch 79/100
15624/15625 [============================>.] - ETA: 0s - loss: 4.2350 - accuracy: 0.2214
Epoch 79: loss improved from 4.23977 to 4.23497, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000079


15625/15625 [==============================] - 136s 9ms/step - loss: 4.2350 - accuracy: 0.2214
Epoch 80/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.2362 - accuracy: 0.2222
Epoch 80: loss did not improve from 4.23497
15625/15625 [==============================] - 123s 8ms/step - loss: 4.2361 - accuracy: 0.2222
Epoch 81/100
15625/15625 [==============================] - ETA: 0s - loss: 4.2312 - accuracy: 0.2227
Epoch 81: loss improved from 4.23497 to 4.23118, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000081


15625/15625 [==============================] - 133s 9ms/step - loss: 4.2312 - accuracy: 0.2227
Epoch 82/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.2322 - accuracy: 0.2224
Epoch 82: loss did not improve from 4.23118
15625/15625 [==============================] - 123s 8ms/step - loss: 4.2323 - accuracy: 0.2224
Epoch 83/100
15624/15625 [============================>.] - ETA: 0s - loss: 4.2354 - accuracy: 0.2226
Epoch 83: loss did not improve from 4.23118
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2354 - accuracy: 0.2226
Epoch 84/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.2300 - accuracy: 0.2231
Epoch 84: loss improved from 4.23118 to 4.22994, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000084


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2299 - accuracy: 0.2231
Epoch 85/100
15623/15625 [============================>.] - ETA: 0s - loss: 4.2246 - accuracy: 0.2238
Epoch 85: loss improved from 4.22994 to 4.22462, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000085


15625/15625 [==============================] - 133s 8ms/step - loss: 4.2246 - accuracy: 0.2238
Epoch 86/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.2268 - accuracy: 0.2233
Epoch 86: loss did not improve from 4.22462
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2266 - accuracy: 0.2234
Epoch 87/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.2250 - accuracy: 0.2233
Epoch 87: loss did not improve from 4.22462
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2248 - accuracy: 0.2233
Epoch 88/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.2221 - accuracy: 0.2237
Epoch 88: loss improved from 4.22462 to 4.22219, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000088


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2222 - accuracy: 0.2237
Epoch 89/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.2254 - accuracy: 0.2235
Epoch 89: loss did not improve from 4.22219
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2254 - accuracy: 0.2236
Epoch 90/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.2206 - accuracy: 0.2236
Epoch 90: loss improved from 4.22219 to 4.22052, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000090


15625/15625 [==============================] - 133s 9ms/step - loss: 4.2205 - accuracy: 0.2236
Epoch 91/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.2241 - accuracy: 0.2241
Epoch 91: loss did not improve from 4.22052
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2240 - accuracy: 0.2241
Epoch 92/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.2162 - accuracy: 0.2241
Epoch 92: loss improved from 4.22052 to 4.21619, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000092


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2162 - accuracy: 0.2240
Epoch 93/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.2151 - accuracy: 0.2241
Epoch 93: loss improved from 4.21619 to 4.21508, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000093


15625/15625 [==============================] - 133s 8ms/step - loss: 4.2151 - accuracy: 0.2241
Epoch 94/100
15623/15625 [============================>.] - ETA: 0s - loss: 4.2139 - accuracy: 0.2246
Epoch 94: loss improved from 4.21508 to 4.21388, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000094


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2139 - accuracy: 0.2246
Epoch 95/100
15623/15625 [============================>.] - ETA: 0s - loss: 4.2190 - accuracy: 0.2244
Epoch 95: loss did not improve from 4.21388
15625/15625 [==============================] - 123s 8ms/step - loss: 4.2190 - accuracy: 0.2244
Epoch 96/100
15622/15625 [============================>.] - ETA: 0s - loss: 4.2141 - accuracy: 0.2244
Epoch 96: loss did not improve from 4.21388
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2141 - accuracy: 0.2244
Epoch 97/100
15619/15625 [============================>.] - ETA: 0s - loss: 4.2123 - accuracy: 0.2244
Epoch 97: loss improved from 4.21388 to 4.21238, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000097


15625/15625 [==============================] - 135s 9ms/step - loss: 4.2124 - accuracy: 0.2244
Epoch 98/100
15621/15625 [============================>.] - ETA: 0s - loss: 4.2124 - accuracy: 0.2251
Epoch 98: loss did not improve from 4.21238
15625/15625 [==============================] - 123s 8ms/step - loss: 4.2124 - accuracy: 0.2251
Epoch 99/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.2085 - accuracy: 0.2255
Epoch 99: loss improved from 4.21238 to 4.20856, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000099


15625/15625 [==============================] - 133s 9ms/step - loss: 4.2086 - accuracy: 0.2255
Epoch 100/100
15620/15625 [============================>.] - ETA: 0s - loss: 4.2060 - accuracy: 0.2254
Epoch 100: loss improved from 4.20856 to 4.20606, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000100


15625/15625 [==============================] - 133s 8ms/step - loss: 4.2061 - accuracy: 0.2254


In [38]:
#model loaded from the previous training. Training is resumed 
model = load_model('/content/drive/MyDrive/Deep Learning/task3/model_weights.h5')
# Load the checkpoints
weight_file = "/content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000101"
model.load_weights(weight_file).expect_partial()

In [ ]:
#Resuming the model training by loading the previous weights
model.fit(X,y, epochs=150, batch_size = 32 ,initial_epoch=101, callbacks=[es, model_checkpoint_callback])

Epoch 102/150
15624/15625 [============================>.] - ETA: 0s - loss: 4.2095 - accuracy: 0.2252
Epoch 102: loss improved from inf to 4.20946, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000102


15625/15625 [==============================] - 156s 9ms/step - loss: 4.2095 - accuracy: 0.2252
Epoch 103/150
15624/15625 [============================>.] - ETA: 0s - loss: 4.2042 - accuracy: 0.2259
Epoch 103: loss improved from 4.20946 to 4.20423, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000103


15625/15625 [==============================] - 133s 8ms/step - loss: 4.2042 - accuracy: 0.2259
Epoch 104/150
15621/15625 [============================>.] - ETA: 0s - loss: 4.2066 - accuracy: 0.2255
Epoch 104: loss did not improve from 4.20423
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2067 - accuracy: 0.2255
Epoch 105/150
15625/15625 [==============================] - ETA: 0s - loss: 4.2069 - accuracy: 0.2259
Epoch 105: loss did not improve from 4.20423
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2069 - accuracy: 0.2259
Epoch 106/150
15619/15625 [============================>.] - ETA: 0s - loss: 4.2102 - accuracy: 0.2254
Epoch 106: loss did not improve from 4.20423
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2102 - accuracy: 0.2254
Epoch 107/150
15621/15625 [============================>.] - ETA: 0s - loss: 4.2082 - accuracy: 0.2262
Epoch 107: loss did not improve from 4.20423
15625/15625 [===============

15625/15625 [==============================] - 130s 8ms/step - loss: 4.2034 - accuracy: 0.2259
Epoch 109/150
15625/15625 [==============================] - ETA: 0s - loss: 4.2037 - accuracy: 0.2264
Epoch 109: loss did not improve from 4.20342
15625/15625 [==============================] - 122s 8ms/step - loss: 4.2037 - accuracy: 0.2264
Epoch 110/150
15623/15625 [============================>.] - ETA: 0s - loss: 4.1984 - accuracy: 0.2259
Epoch 110: loss improved from 4.20342 to 4.19850, saving model to /content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000110


15625/15625 [==============================] - 133s 8ms/step - loss: 4.1985 - accuracy: 0.2259
Epoch 111/150
15620/15625 [============================>.] - ETA: 0s - loss: 4.2007 - accuracy: 0.2269
Epoch 111: loss did not improve from 4.19850
15625/15625 [==============================] - 121s 8ms/step - loss: 4.2007 - accuracy: 0.2269
Epoch 112/150
15622/15625 [============================>.] - ETA: 0s - loss: 4.1994 - accuracy: 0.2265
Epoch 112: loss did not improve from 4.19850
15625/15625 [==============================] - 121s 8ms/step - loss: 4.1993 - accuracy: 0.2265
Epoch 113/150
15620/15625 [============================>.] - ETA: 0s - loss: 4.2017 - accuracy: 0.2265
Epoch 113: loss did not improve from 4.19850
15625/15625 [==============================] - 121s 8ms/step - loss: 4.2016 - accuracy: 0.2265
Epoch 114/150
15619/15625 [============================>.] - ETA: 0s - loss: 4.2016 - accuracy: 0.2266
Epoch 114: loss did not improve from 4.19850
15625/15625 [===============

The loss did not imrpove after 110 epochs where the training is terminated as the patience set in early stopping is set to 0

In [47]:
#model loaded from the previous training. Training is resumed 
model = load_model('/content/drive/MyDrive/Deep Learning/task3/model_weights.h5')
# Load the checkpoints
weight_file = "/content/drive/MyDrive/Deep Learning/task3/checkpoints/task3_checkpoints.0000110"
model.load_weights(weight_file).expect_partial()

In [40]:
# function to generate text when input text and number of words to be generated are given. The text is preprocessed. 
def text_generator(model, tokenizer, seq_len, feature_text, num_words):
  text = []
  for i in range(num_words):
    token = tokenizer.texts_to_sequences([feature_text])[0]
    token = pad_sequences([token], maxlen = seq_len, truncating='pre')
    # y_pred = model.predict_classes(token)
    y_pred = model.predict(token) 
    y_pred = np.argmax(y_pred, axis=1)

    pred_word = ''
    for word, idx in tokenizer.word_index.items():
      if idx == y_pred:
        pred_word = word
        break
    feature_text += " "+ pred_word
    text.append(pred_word)

  return " ".join(text)

In [46]:
while True:
    text = input('Enter your line: ')
    if text.lower() == 'q':
        print('Session terminated...')
        break

    num_of_words = int(input('Enter the number of words to generate: '))

    try:
        print(text)

        next_sentences = text_generator(model, tokenizer, X.shape[1], text, num_of_words)
        print("The next sentences are:", next_sentences)
        print("Enter q to quit")

    except Exception as e:
        print('Error occurred:', e)
        continue


Enter your line: sea monsters are coming to get us and we have to get back to the shore as soon as possible
Enter the number of words to generate: 50
sea monsters are coming to get us and we have to get back to the shore as soon as possible
1/1 [==============================] - 0s 36ms/step
The next sentences are: astronomers to allow nine down in the sun whose tongue is written in the cells the captain is the same thing the chief scientist came together marin had grown forever weapons is hardly the final landing the golden section says appeared for the extreme stage and successful supply in the
Enter q to quit
Enter your line: q
Session terminated...
